观察:`QueryParams.$filter` 声明为 `Record< string, any >`,但仓内下沉口早已往里送 ObjectQL AST 数组

观察类发现(`finding`,不进 `pm:queue`),发现于 #3831 / PR #3908 的实施过程。运行时正确、今天没有用户会撞到 —— 是类型面与实际契约的漂移。

## 机制

`packages/types/src/data.ts:53` 声明:

```
/**
 * Filter expression
 * @example { age: { $gt: 18 }, status: 'active' }
 */
$filter?: Record< string, any >;
```

即只描述 MongoDB 风格的 field-keyed 记录形式。但仓内规范下沉口 `packages/core/src/utils/filter-converter.ts` 的 `toFilterNode` / `mergeFilterNodes` 返回的是 `FilterNode | Record< string, any > | undefined` —— **AST 数组是其正常返回值之一**,而两个既有消费者一直把它送进 `$filter`:

- `packages/plugin-list/src/ListView.tsx` 的 `buildEffectiveFilter(baseFilter: unknown, …)`,返回 `any[] | Record< string, any > | undefined`(喂表格与导出两条路);
- `packages/plugin-view/src/ObjectView.tsx`(calendar / kanban / gallery / timeline)。

数据源侧完全支持:`packages/data-objectstack/src/index.ts` 的 `translateFilterToAST` 明确列出五种可接受输入,数组形式在列。所以**运行时是对的,声明是窄的**。

## 有害在哪(只有类型面)

1. 该类型不设防且**方向错误**:`Record< string, any >` 会接受数组(数组满足 `any` 的字符串索引 —— `tsc --strict` 实测 exit 0),所以它既没挡住不该进的形状,也没描述该进的形状。#3831 正是踩在这一点上:`baseFilter?: Record< string, any >` 接受了规则数组、随后被对象展开压成 `{"0": {...}}`,类型全绿而查询去过滤名为 `0` 的列。
2. 新写消费者的人读 `$filter` 的类型与 `@example`,会以为只有记录形式合法,于是可能给数组路径补一段"容错转换" —— 正是 AGENTS.md #0.1 反对的方向。

## 落地建议

把 `$filter` 声明成它实际接受的联合(记录形式 | AST 节点),`@example` 补一条数组示例,并在注释里指名 `translateFilterToAST` 为权威可接受集。属 `packages/types` 公共面,会波及若干消费者的类型标注,故不宜搭在功能 PR 上 —— PR #3908 只在单个赋值点(`packages/fields/src/widgets/useRecordQuery.ts`)就地转换并注明原因,没有顺手改这个共享类型。

## 参考位置

- `packages/types/src/data.ts:53`(声明)
- `packages/core/src/utils/filter-converter.ts`(`toFilterNode` / `mergeFilterNodes` 的返回类型)
- `packages/plugin-list/src/ListView.tsx`(`buildEffectiveFilter`)/ `packages/plugin-view/src/ObjectView.tsx`(两个既送数组的生产者)
- `packages/data-objectstack/src/index.ts`(`translateFilterToAST`,实际可接受集)

关联:#3831 / PR #3908(发现来源)、#3948(两份 operator 词汇表漂移的先例)

os-dev-report

```json
{
  "issue": 3909,
  "status": "done",
  "branch": "claude/issue-3909-queryparams-filter-union",
  "pr": "https://github.com/objectstack-ai/objectui/pull/5999",
  "premise_still_valid": true,
  "summary": "Every premise re-derived on origin/main @ d8b48f495 and all still hold, including the line number: packages/types/src/data.ts:53 still declared $filter as a record with a record-only @example; filter-converter.ts's toFilterNode/mergeFilterNodes (:285/:323) still return FilterNode-or-record; ListView's buildEffectiveFilter (:470) and ObjectView (:730/:778) still feed arrays through the slot; translateFilterToAST (data-objectstack/src/index.ts:306) still enumerates five accepted inputs. Declared $filter as `Record<string, any> | FilterArray`, BINDING the array half to @objectstack/spec/data's FilterArray rather than re-declaring a local FilterNode copy -- this is the file's own stated convention ('Spec-owned names are bound here, not re-declared') and it avoids creating the third partial list to drift from that the card warns about; the spec's own FilterArray doc names 'the wire $filter face' as a measured producer, so the slot was already in its declared scope. Added array @examples and named translateFilterToAST in the comment as the authoritative accepted set. Also removed the PR #3908 workaround (declared below). Type-only; zero runtime change.",
  "tests": "All at final commit ab70c650. BUILD/GATES: `pnpm --filter @object-ui/types build` exit 0; `pnpm --filter @object-ui/types type-check` exit 0 (script echoed `tsc --noEmit && tsc -p tsconfig.examples.json && tsc -p tsconfig.test.json`, so not the zero-match trap); `pnpm --filter @object-ui/fields type-check` exit 0 (script echoed); `pnpm exec vitest run packages/types/src/__tests__/query-params-filter-union.test.ts` -> 'Tests 6 passed (6)'. Exit codes captured by redirect-then-capture, never after a pipe. DOWNSTREAM SWEEP (PM assumption 2): `turbo run build --filter '...@object-ui/types'` -- PREFIX filter, i.e. DEPENDENTS/downstream, not dependencies -- printed 'Tasks: 43 successful, 43 total' at HEAD ab70c650. CLAUSE-2 RE-MEASUREMENT: confirmed with the repo's own tsc 6.0.3 that Record<string,any> already structurally accepts arrays -- all five translateFilterToAST input shapes assign with NO cast under both the old and new declaration (exit 0 both), and a negative control proves the probe can fail: `$filter: 42` and `$filter: 'status eq active'` are rejected identically by both (exit 2, TS2322 x4), which also proves QueryParams' `[key: string]: any` index signature is not swallowing the slot. Accepted set unchanged, so Clause-2 stays 'no' and this is not a narrowing. ABLATION / REVERSE VERIFICATION -- and one correction worth flagging: my FIRST reverse-verification reading was VOID and I am reporting it rather than quietly re-running. I set a restoring `trap ... EXIT` in one bash call and measured in the NEXT call, so the trap had already restored data.ts and I measured the unmutated tree; it read green and would have been recorded as 'the pin does not fire'. Redone with mutation and measurement inside ONE invocation, mutation confirmed on disk by grepping both the injected and the removed text (injected:1, removed-remaining:0) rather than trusting an editor exit code. That real run exposed a genuine defect in my own first pin: it was a PHANTOM. Reverting $filter to the bare record left type-check GREEN (exit 0), because assignability cannot separate the two -- arrays satisfy Record<string,any>'s string index, so `FilterArray extends QueryParams['$filter']` holds under BOTH declarations and the old declaration is itself assignable to the new union. Replaced it with an exact-identity assert (`Equal<NonNullable<QueryParams['$filter']>, Record<string,any> | FilterArray>`), which also catches the subtler regression of a local AST fork replacing the spec binding. Re-verified from a COMMITTED state: baseline type-check green exit 0; mutated type-check RED exit 2 with `src/__tests__/query-params-filter-union.test.ts(112,7): error TS2344: Type 'false' does not satisfy the constraint 'true'`; mutated vitest GREEN 6/6 -- the runtime-blind/compile-red split reproduced deliberately, which is this drift's own signature. Trap restored the tree and `git diff` on data.ts is empty. LINT: ran the full repo-wide `eslint . --no-inline-config --format json` (it fits inside the foreground cap, so no narrowing was needed): population 3626 files read from eslint's own config, 74 error-bearing files, and NONE of them are mine -- my three files contribute 0 errors and 37 warnings, all pre-existing no-explicit-any/react-hooks on lines this diff does not touch. Type-aware linting is not configured (no projectService/parserOptions.project in eslint.config.js), so this diff provably cannot move any untouched file's verdict. CONTROL PROBE for the path re-derivation: a real symbol greps exit 0 and a fabricated one exits 1, captured as grep's own status rather than through a pipe, so the zero-hits read as real. NOT RUN LOCALLY: the rest of the CI farm, per the report-at-draft-PR contract; CI convergence is the PM's read.",
  "open_questions": [],
  "out_of_scope_findings": [
    "filed as #6001: ValueDataSource.test.ts carries 10 `$filter: [...] as any` casts that exist only because the type did not describe arrays; they are redundant after this change, and the `as any` erases the type-face evidence those nine AST shapes would otherwise provide (a narrowing back to the record form is invisible to that file). Filed unassigned with the `finding` label, no pm:queue, Blocked-by #3909; deliberately NOT fixed here -- different package, and it would widen this PR's verification surface beyond the fence."
  ]
}
```

## PM assumption 3 — the #3908 workaround, declared either way

**It became removable, and I removed it.** `packages/fields/src/widgets/useRecordQuery.ts:180` held `params.$filter = filter as Record<string, any>` with a comment stating outright that the cast sat at that one assignment *because* the shared type had not been widened. Once the shared type is honest that comment is a false statement about the tree, so leaving it would have been worse than the cast.

The cast could not simply be deleted — `filter` is declared `unknown`, and `hasFilter` was a plain `boolean`, so nothing narrowed it. `hasFilter` is now a type predicate narrowing to the slot's **own** type (`filter is NonNullable<QueryParams['$filter']>`) rather than a restatement of it, so the assignment needs no cast and the guard cannot drift from the declaration it guards. The predicate is sound: it returns true only for a non-empty array or a non-empty object, and both are assignable to the union. `@object-ui/fields` type-check exit 0, and the package is in the green 43/43 sweep.

Scope note: this is the only non-test cast in the repo caused by this defect. I swept all `$filter` assignment sites; the one other non-test `as any` (`data-objectstack/src/index.ts:4455`) is an unrelated `params.filter` shape and I left it alone.

---
_Generated by [Claude Code](https://claude.ai/code)_

--- a/.changeset/3909-query-params-filter-union.md
+++ b/.changeset/3909-query-params-filter-union.md
@@ -1,0 +1,38 @@
+---
+'@object-ui/types': patch
+'@object-ui/fields': patch
+---
+
+`QueryParams.$filter` now declares both shapes the data sources actually accept — the
+MongoDB-style field-keyed record, or a `FilterArray`, the ObjectQL AST sugar bound from
+`@objectstack/spec/data` (objectui#3909).
+
+**Nothing is narrowed and no accepted value changes.** `Record<string, any>` already
+accepted arrays structurally — they satisfy its string index — so the union documents
+shapes that were always legal rather than admitting new ones. Measured both ways under
+`tsc --strict`: all five inputs `translateFilterToAST` enumerates assign to the old and
+new declarations alike, and both reject a bare number and a bare string identically. A
+downstream `turbo run build` over all 43 dependent packages is green, which is the
+evidence a published type change breaks no consumer.
+
+The harm was entirely on the type face, and it was two-sided. The declaration blocked
+nothing while describing one legal shape as though it were the only one — objectui#3831
+is what that cost, a rule array accepted by a `Record<string, any>` slot, object-spread
+flattened to `{"0": {...}}`, types green, and the query filtering on a column literally
+named `0`. And someone writing a new consumer would read the type and its record-only
+`@example`, conclude the array path was illegal, and add a tolerant conversion for it —
+the "widen the consumer to tolerate the producer" shape AGENTS.md #0.1 forbids. Two
+producers have fed arrays through this slot all along: `plugin-list`'s
+`buildEffectiveFilter` (grid and export) and `plugin-view`'s `ObjectView` (calendar /
+kanban / gallery / timeline). The runtime was right; the declaration was narrow.
+
+The array half is **bound** to the spec's `FilterArray` rather than restated locally, so
+it cannot fork from the vocabulary the servers parse — the same failure two hand-written
+operator lists had in objectui#3948. The doc comment names `translateFilterToAST` as the
+authoritative accepted set instead of carrying a second list to drift from.
+
+`@object-ui/fields` drops the local cast this defect forced. PR objectui#3908 wrote
+`filter as Record<string, any>` at one assignment in `useRecordQuery`, deliberately, as
+debt rather than widening the shared type. `hasFilter` is now a type predicate narrowing
+to the `$filter` slot's own type, so the assignment needs no cast and the guard cannot
+drift from the declaration it guards. Type-only throughout; no runtime behaviour changes.

--- a/packages/fields/src/widgets/useRecordQuery.ts
+++ b/packages/fields/src/widgets/useRecordQuery.ts
@@ -111,8 +111,12 @@ export interface UseRecordQueryResult {
  * `Object.keys` on an array returns its INDICES — so the record-only test read
  * `['0','1','2']` for an AST node and was right only by accident. An empty array
  * is "no filter" for the same reason an empty object is.
+ *
+ * Narrows to the `$filter` slot's own type rather than a restatement of it
+ * (#3909), so the caller assigns with no cast and this predicate cannot drift
+ * from the declaration it is guarding.
  */
-function hasFilter(filter: unknown): boolean {
+function hasFilter(filter: unknown): filter is NonNullable<QueryParams['$filter']> {
   if (filter === null || filter === undefined) return false;
   if (Array.isArray(filter)) return filter.length > 0;
   if (typeof filter !== 'object') return false;
@@ -172,12 +176,14 @@ export function useRecordQuery(options: UseRecordQueryOptions): UseRecordQueryRe
         if (searchTerm && searchTerm.trim()) params.$search = searchTerm.trim();
         if (searchFields && searchFields.length > 0) params.$searchFields = searchFields;
         if (sortArg) params.$orderby = { [sortArg.field]: sortArg.direction };
-        // `QueryParams.$filter` is declared `Record< string, any >`, which the
-        // AST-array form does not describe — the cast is at this ONE assignment
-        // rather than widening a shared type that several other producers
-        // (plugin-list's `buildEffectiveFilter`, plugin-view's ObjectView)
-        // already feed arrays through.
-        if (hasFilter(filter)) params.$filter = filter as Record<string, any>;
+        // No cast: `QueryParams.$filter` now declares both shapes it accepts
+        // (#3909), so the AST-array form the picker's merge yields is describable
+        // here. The local cast this replaces was deliberate debt — taken at this
+        // ONE assignment rather than widening the shared type that several other
+        // producers (plugin-list's `buildEffectiveFilter`, plugin-view's
+        // ObjectView) already feed arrays through. The shared type is honest now,
+        // so the debt is paid rather than moved.
+        if (hasFilter(filter)) params.$filter = filter;
         if (expand && expand.length > 0) params.$expand = expand;
 
         const result = await dataSource.find(objectName, params);

--- a/packages/types/src/__tests__/query-params-filter-union.test.ts
+++ b/packages/types/src/__tests__/query-params-filter-union.test.ts
@@ -1,0 +1,112 @@
+/**
+ * ObjectUI
+ * Copyright (c) 2024-present ObjectStack Inc.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+/**
+ * objectui#3909 — `QueryParams.$filter` declares BOTH shapes the data sources
+ * accept, and binds the array half to `@objectstack/spec`'s `FilterArray`
+ * rather than restating it.
+ *
+ * ## Why this file exists at all
+ *
+ * The defect it pins was invisible to every runtime suite, and had to be: the
+ * declaration was `Record< string, any >`, which **structurally accepts arrays**
+ * (they satisfy its string index). So the two producers that have fed ObjectQL
+ * AST arrays through this slot all along — `plugin-list`'s
+ * `buildEffectiveFilter` (grid and export) and `plugin-view`'s `ObjectView`
+ * (calendar / kanban / gallery / timeline) — type-checked, ran, and shipped
+ * correct results. Nothing was broken at runtime and nothing could go red.
+ *
+ * The cost was paid on the type face instead, in both directions:
+ *
+ * 1. The type **blocked nothing while describing one legal shape as if it were
+ *    the only one**. objectui#3831 is what that buys: a `Record< string, any >`
+ *    slot accepted a rule array, an object spread flattened it to
+ *    `{"0": {...}}`, types stayed green, and the query filtered on a column
+ *    literally named `0`.
+ * 2. Someone writing a new consumer reads the type and its `@example`,
+ *    concludes only the record form is legal, and adds a tolerant conversion
+ *    for the array path — the "widen the consumer to tolerate the producer"
+ *    shape AGENTS.md #0.1 forbids.
+ *
+ * Both failure modes are compile-time by nature, so the pins are too. Reverting
+ * `$filter` to `Record< string, any >` leaves every runtime suite green and
+ * turns THIS FILE red under `tsc -p tsconfig.test.json` (the `type-check`
+ * script) — the drift's own signature, reproduced deliberately.
+ *
+ * ## What is NOT pinned here
+ *
+ * That the union is the *authoritative* accepted set. It is not: the authority
+ * is `translateFilterToAST` (`@object-ui/data-objectstack`), which enumerates
+ * five input shapes. A second list here would be a third place to drift from —
+ * which is exactly how two operator vocabularies came apart in #3948. The
+ * binding below is to the spec's `FilterArray`, so the array half cannot fork
+ * locally.
+ */
+
+import { describe, it, expect } from 'vitest';
+import type { FilterArray } from '@objectstack/spec/data';
+import type { QueryParams } from '../data';
+
+type Assert< T extends true > = T;
+/** True when `V` is accepted by the `$filter` slot. */
+type AcceptsFilter< V > = V extends QueryParams['$filter'] ? true : false;
+
+describe('QueryParams.$filter — declares the union it actually accepts (#3909)', () => {
+  it('accepts the MongoDB-style field-keyed record', () => {
+    type _Record = Assert< AcceptsFilter< { age: { $gt: number } } > >;
+    const params: QueryParams = { $filter: { age: { $gt: 18 }, status: 'active' } };
+    expect(params.$filter).toEqual({ age: { $gt: 18 }, status: 'active' });
+  });
+
+  it('accepts a bare AST comparison tuple with no cast', () => {
+    // The shape `buildEffectiveFilter` returns for a single condition. Before
+    // #3909 this compiled only because arrays satisfy `Record`'s string index —
+    // accepted by accident rather than by declaration.
+    const params: QueryParams = { $filter: ['status', '=', 'active'] };
+    expect(params.$filter).toEqual(['status', '=', 'active']);
+  });
+
+  it('accepts a logical AST group with no cast', () => {
+    // What `mergeFilterNodes` returns once more than one source is active.
+    const params: QueryParams = {
+      $filter: ['and', ['age', '>=', 18], ['status', '=', 'active']],
+    };
+    expect(Array.isArray(params.$filter)).toBe(true);
+  });
+
+  it('accepts the legacy bare list, combined with implicit AND', () => {
+    const params: QueryParams = {
+      $filter: [['stage', '=', 'won'], ['amount', '>', 1000]],
+    };
+    expect(Array.isArray(params.$filter)).toBe(true);
+  });
+
+  it('binds the array half to the spec rather than restating it', () => {
+    // A locally re-declared AST type would satisfy the assignments above just
+    // as well — and would then be free to drift from the spec's vocabulary the
+    // way two hand-written operator lists did (#3948). This pin fails if the
+    // union stops admitting the spec's own `FilterArray`.
+    type _Bound = Assert< AcceptsFilter< FilterArray > >;
+    const fromSpec: FilterArray = ['status', '=', 'active'];
+    const params: QueryParams = { $filter: fromSpec };
+    expect(params.$filter).toBe(fromSpec);
+  });
+
+  it('still refuses a value that is neither shape', () => {
+    // The union documents; it must not have become `any` on the way. Note the
+    // slot sits on an interface that also carries `[key: string]: any` — these
+    // pins prove the declared property still wins over that index signature,
+    // which is the whole reason the declaration is worth anything.
+    // @ts-expect-error a number is not a filter
+    const bad: QueryParams = { $filter: 42 };
+    // @ts-expect-error a string is not a filter
+    const alsoBad: QueryParams = { $filter: 'status eq active' };
+    expect(bad.$filter).toBe(42);
+    expect(alsoBad.$filter).toBe('status eq active');
+  });
+});

--- a/packages/types/src/__tests__/query-params-filter-union.test.ts
+++ b/packages/types/src/__tests__/query-params-filter-union.test.ts
@@ -55,6 +55,9 @@ import type { QueryParams } from '../data';
 type Assert< T extends true > = T;
 /** True when `V` is accepted by the `$filter` slot. */
 type AcceptsFilter< V > = V extends QueryParams['$filter'] ? true : false;
+/** Exact type identity — NOT mutual assignability. See the note below. */
+type Equal< A, B > =
+  (< T >() => T extends A ? 1 : 2) extends (< T >() => T extends B ? 1 : 2) ? true : false;
 
 describe('QueryParams.$filter — declares the union it actually accepts (#3909)', () => {
   it('accepts the MongoDB-style field-keyed record', () => {
@@ -86,12 +89,28 @@ describe('QueryParams.$filter — declares the union it actually accepts (#3909)
     expect(Array.isArray(params.$filter)).toBe(true);
   });
 
-  it('binds the array half to the spec rather than restating it', () => {
-    // A locally re-declared AST type would satisfy the assignments above just
-    // as well — and would then be free to drift from the spec's vocabulary the
-    // way two hand-written operator lists did (#3948). This pin fails if the
-    // union stops admitting the spec's own `FilterArray`.
-    type _Bound = Assert< AcceptsFilter< FilterArray > >;
+  it('binds the array half to the spec, by IDENTITY not assignability', () => {
+    // ## Why this pin is an identity check, and why nothing weaker works
+    //
+    // Every assignment-shaped pin in this file is, on its own, VACUOUS as a
+    // regression guard — measured, not assumed. Reverting the declaration to
+    // `Record< string, any >` and re-running `type-check` leaves it GREEN
+    // (exit 0), because assignability cannot separate the two: arrays satisfy
+    // `Record< string, any >`'s string index, so `FilterArray extends
+    // QueryParams['$filter']` holds under BOTH declarations, and the old
+    // declaration is itself assignable to the new union. A guard that passes
+    // equally before and after the fix is a phantom check — it reads like
+    // enforcement and enforces nothing.
+    //
+    // Identity is the property that actually differs. This assertion goes red
+    // on a revert to the bare record, AND on the subtler regression: someone
+    // re-declaring a local `FilterNode` fork instead of binding the spec's
+    // type. That fork would satisfy every assignment above while being free to
+    // drift from the vocabulary the servers parse — the exact failure two
+    // hand-written operator lists had in #3948.
+    type _Bound = Assert<
+      Equal< NonNullable< QueryParams['$filter'] >, Record< string, any > | FilterArray >
+    >;
     const fromSpec: FilterArray = ['status', '=', 'active'];
     const params: QueryParams = { $filter: fromSpec };
     expect(params.$filter).toBe(fromSpec);

--- a/packages/types/src/data.ts
+++ b/packages/types/src/data.ts
@@ -31,6 +31,7 @@ import type {
   CreateExportJobInput as SpecCreateExportJobInput,
   CreateExportJobResult,
 } from '@objectstack/spec/contracts';
+import type { FilterArray } from '@objectstack/spec/data';
 import type { ValidationError } from '@objectstack/spec/kernel';
 
 export type { ExportJobStatus, ImportJobStatus, ImportWriteMode, ValidationError };
@@ -47,10 +48,39 @@ export interface QueryParams {
   $select?: string[];
 
   /**
-   * Filter expression
+   * Filter expression, in either of the two forms the data sources accept:
+   * the MongoDB-style field-keyed record, or a `FilterArray` — the spec-owned
+   * ObjectQL AST sugar (`@objectstack/spec/data`).
+   *
+   * Both forms are normal here, and the array form is not an edge case: the
+   * repo's own canonical sink `mergeFilterNodes` / `toFilterNode`
+   * (`@object-ui/core`'s `filter-converter.ts`) returns AST nodes, and its two
+   * standing producers — `plugin-list`'s `buildEffectiveFilter` (grid and
+   * export) and `plugin-view`'s `ObjectView` (calendar / kanban / gallery /
+   * timeline) — have fed arrays through this slot all along.
+   *
+   * ⛔ Do not add a "tolerant conversion" in a consumer to cope with the array
+   * path. The array IS legal input; a consumer that needs one shape lowers
+   * through the shared sink rather than widening itself to tolerate the
+   * producer.
+   *
+   * The authoritative acceptable set is the one `translateFilterToAST`
+   * (`@object-ui/data-objectstack`'s `index.ts`) enumerates — five input
+   * shapes, of which the array forms below are three. Read it there rather than
+   * trusting a second list here; a partial restatement is exactly how two
+   * operator vocabularies drifted apart before.
+   *
+   * Note the declaration does not *narrow* anything: `Record<string, any>`
+   * already structurally accepts arrays (they satisfy its string index), so the
+   * union documents the shapes that were always accepted rather than admitting
+   * new ones. It is the description that was wrong, not the runtime.
+   *
    * @example { age: { $gt: 18 }, status: 'active' }
+   * @example ['status', '=', 'active']
+   * @example ['and', ['age', '>=', 18], ['status', '=', 'active']]
+   * @example [['stage', '=', 'won'], ['amount', '>', 1000]]
    */
-  $filter?: Record<string, any>;
+  $filter?: Record<string, any> | FilterArray;
 
   /**
    * Sort order

